File: Binary_Search/acw_790-3sqrt_float_bin_search.py
# why 1e-8?(↓经验：误差缩小2位)
# -- 这样的话可以把误差缩小到1e-8，题目要求只输出6位小数
def float_bsearch(num, l, r):
    def q(num):
        return num * num * num

    while l + 1e-8 <= r:
        mid = (l + r) / 2  # 浮点数
        if q(mid) > num:
            r = mid
        else:
            l = mid
    return l


if __name__ == "__main__":
    # ∵−10000≤n≤10000 ∴3次方根∈[-22, 22]
    n = float(input())
    l, r = -22, 22
    ans = float_bsearch(n, l, r)
    print(f'{ans:.6f}')

File: Binary_Search/acw_790_3sqrt_float_bin_search.java
package Binary_Search;

import java.util.Scanner;

public class acw_790_3sqrt_float_bin_search {
    private static final double EPS = 1e-8;
    private static double binSearch_float(double x, double start, double end) {
        while (start + EPS < end) { // [L, mid] [mid, R]
            double mid = (start + end) / 2.0;
            if (cube(mid) < x) {
                start = mid; // 无需+1！double型也不应该加的是1.0！
            } else end = mid;
        }
        return start;
    }

    private static double cube(double num) {
        return num * num * num;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        double n = sc.nextDouble();
        // # ∵−10000≤n≤10000 ∴3次方根∈[-22, 22]
        double start = -22.0, end = 22.0;
        double ans = binSearch_float(n, start, end);
        System.out.format("%.6f", ans);
    }
}